enhancements/substitutes-for.md

```diff
@@ -0,0 +1,379 @@
+
+Note also that each time a new rule is added, the table must be re-processed for additional
+new rules, until no new rules are generated.
+
+Alternatively the transitive substitution relationships could be derived at runtime when
+ListBundles is performed, following the substitution chains of any "skips" and "replaces"
+references to add the additional skips entries.
+
+Given the set of substitution relationships:
+A' -> A
+A'' -> A'
+A'' -> A
+
+As well as the replaces definition in the CSV (B replaces A), it remains for ListBundles
+to produce a CSV that will allow OLM to upgrade versions A'', A', and A' of the operator, 
+to version B.
+
+To achieve this, the CSV must be mutated as part of the ListBundles result.  Since
+"replaces" can only have a single entry, we will need to add additional "skips" entries
+to the CSV to account for the other valid upgrade paths.  Specifically, the mutated
+CSV for "B" will indicate that it skips A'' and A' (it already "replaces" A')
+
+In addition, the mutated version of A'' will add a skip for A.  (A'' will already skip A' as 
+provided by freshmaker).  Finally, A' will also skip A, as defined by freshmaker.
+
+This means the CSVs, as seen by OLM, will be:
+
+B: replaces A(original CSV), skips A', A''(result of substitution of A)
+A: whatever the original CSV listed
+A': skips A(added by freshmaker), plus original CSV
+A'': skips A'(added by freshmaker)+A(carried from CSV for A' when A' was rebuilt)+original CSV
+
+### User Stories [optional]
+
+Detail the things that people will be able to do if this is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+
+#### Freshmaker rebuilds a bundle that is the head of the channel, the bundle
+does not exist in any other channels.
+
+In this case, freshmaker simply takes the CSV it is rebuilding and adds a
+"skips" reference naming the bundle it is rebuilding.  A substitutes-for
+reference must also be added to the CSV.
+
+#### Bundle B replaces bundle A, bundle A is rebuilt as A' to fix a CVE.
+
+Freshmaker will rebuild A as A', adding a "skips: A" reference.  In addition
+it will add a "substitutes-for: A" reference.
+
+When ListBundles serves the CSV for B, B will include the following references:
+1) replaces A (from the original CSV)
+2) skips A' (because of replaces A plus the A->A' substitution rule).  Ideally
+this would be a "replaces" reference, but we cannot have multiple replaces
+references.
+
+The result of this is that users on A can update to A' or B(depending which
+channel they are on).  Users on A' can update to B.
+
+**Note: This is not ideal since it means a user can be upgraded from something older
+than A/A', directly to B, without going through A/A', which may mean they
+skip a necessary migration.  Need to discuss this scenario further**
```

Comment:
Is there a reason why, since we're doing mutation on the API call anyway, that we can't flip this? The result would be that B replaces A' instead?

Reply:
we could and i discuss this option under `Implementation Details/Notes/Constraints `, but it is more complicated than not doing it that way.

Reply:
Actually this was discussing a different problem.

If I have the following:
C (oldest)
B, B'' (both replace C)
A (replaces B'', skips B)

And I have C installed, what upgrade path will OLM take and what will be installed on the way there?

Will it upgrade to B' and then to A?
Or will it see that it can "Skip B" and go straight to A without installing B, since A skips B?

And generally how does it choose between B and B''?